File: problem243.scala
import BigInt._

def isPrime(n: Int): Boolean = {
  if (n < 2) false
  else (2 until (math.sqrt(n)).toInt + 1) forall (n % _ != 0)
}

def primeFactors(k:Int, x: Int, factors: Set[Int]): Set[Int] =
  if (x == 1) factors
  else if (x % primes(k) == 0)
    primeFactors(k, x / primes(k), factors + primes(k))
  else primeFactors(k+1, x, factors)

def phiEuler(x: Int, xs: Set[Int]): Int = {
  //println(x, xs)
  def iter(xs: Set[Int], n: BigInt, d: BigInt): Tuple2[BigInt, BigInt] = {
    //println(xs, n, d)
    if (xs.isEmpty) (n, d)
    else iter(xs.tail, n * (xs.head - 1), d * xs.head)
  }
  val r = iter(xs, 1, 1)
  (x * r._1 / r._2).toInt
}

val limit = 20000000
val primes = (2 to limit) filter isPrime
println("Number of primes below ", limit, primes.size)

//val n = 1000003
//println(phiEuler(n, primeFactors(0, n, Set())))

var i = 2500000;
 val lim = 15499.toDouble / 94744
 var r = 0.toDouble
 var phi = 0.toDouble
 do {
 i += 1;
 phi = phiEuler(i, primeFactors(0, i, Set())).toDouble
 r = phi / (i - 1)
 if (i % 10000 == 0) println (i, r)
 } while(r >= lim)
 println(i, phi, r)

File: problem243.cc
#include <stdio.h>
#include <math.h>
#include <vector>
#include <string>
#include <set>

using namespace std;
		   
vector<int> primes;

bool isPrime(int n) {
  for (int i = 2; i < (int)sqrt(n) + 1; ++i) {
    if (n % i == 0) return false;
  }
  return true;
}

set<int> primeFactors(int n) {
  set<int> ret;
  int i = 0;
  int p = 0;
  int size = primes.size();
  while (n > 1 && i < size) {
    p = primes[i];
    if (n % p == 0) {
      n = n / p;
      ret.insert(p);
    } else {
      ++i;
    }
  }
  return ret;
}

int phiEuler(int x) {
  int n = 1;
  int d = 1;
  for (int f : primeFactors(x)) {
    n = n * (f - 1);
    d = d * f;
  }
  return x / d * n;
}

string getPrimes(set<int> primes) {
  string ret = "";
  for (const auto &p : primes) {
    ret += to_string(p);
    ret += " ";
  }
  return ret;
}

void printLine(int i, int minNumber, int minPhi, int n, int d) {
  printf("n:%10d min number: %10d min phi: %10d min value:%12.10f target:%12.10f prime factor: %s\n",
	 i,
	 minNumber,
	 minPhi,
	 (double)minPhi/(minNumber - 1),
	 (double)n/d,
	 getPrimes(primeFactors(minNumber)).c_str()
	 );
}

void bruteForce() {
  int phi = 1;
  int minPhi = 1;
  int minNumber = 2;
  int i = 1;
  int n = 4;
  int d = 10;
  n = 15499;
  d = 94744;
  do {
    ++i;
    phi = phiEuler(i);
    if ((long long)phi * (minNumber - 1) < (long long)minPhi * (i - 1)) {
      minNumber = i;
      minPhi = phi;
    }
    if (i % 1000000 == 0)
      printLine(i, minNumber, minPhi, n, d);
  } while ((long long)phi * d >= (long long)n * (i - 1) /*&& i < 20000000*/); 
  printf("\n");
  printLine(i, minNumber, minPhi, n, d);
}

void setPrimes(int limit) {
  for (int i = 2; i < limit; ++i) {
    if (isPrime(i)) {
      primes.push_back(i);
    }
  }
}

int main() {
  vector<int> primes;
  int limit = 25;
  setPrimes(limit);
  bruteForce();
}
